refactoring.api/src/org/netbeans/modules/refactoring/api/RefactoringSession.java

```diff
@@ -116,79 +108,66 @@ private Problem reallyDoRefactoring(boolean saveAfterDone) {
             try {
                 while (it.hasNext()) {
                     RefactoringElementImplementation element = (RefactoringElementImplementation) it.next();
-                    if (element.isEnabled() && !((element.getStatus() == RefactoringElement.GUARDED) || (element.getStatus() == RefactoringElement.READ_ONLY))) {
-                        element.performChange();
-                    }
+                    changeRefactoringElement(element, true);
                     current += progressStep;
                     fireProgressListenerStep((int) current);
                 }
             } finally {
-                for (Transaction commit : commits) {
-                    SPIAccessor.DEFAULT.check(commit, false);
-                }
-
-                UndoableWrapper wrapper = MimeLookup.getLookup("").lookup(UndoableWrapper.class);
-                for (Transaction commit : commits) {
-                    if (wrapper != null) {
-                        setWrappers(commit, wrapper);
+                commits.forEach(commit -> SPIAccessor.DEFAULT.check(commit, false));
+                UndoableWrapper undoableWrapper = MimeLookup.getLookup("").lookup(UndoableWrapper.class);
+                commits.forEach(commit -> {
+                    if (undoableWrapper != null) {
+                        undoableWrapper.setActive(true, this);
                     }
-
-                    if(commit instanceof ProgressProvider) {
+                    if (commit instanceof ProgressProvider) {
                         ProgressProvider progressProvider = (ProgressProvider) commit;
                         progressProvider.addProgressListener(progressListener);
-                    }
-                    try {
-                        commit.commit();
-                    } finally {
-                        if(commit instanceof ProgressProvider) {
-                            ProgressProvider progressProvider = (ProgressProvider) commit;
+                        try {
+                            commit.commit();
+                        } finally {
                             progressProvider.removeProgressListener(progressListener);
                         }
+                    } else {
+                        commit.commit();
                     }
-                    if (wrapper != null) {
-                        unsetWrappers(commit, wrapper);
+                    if (undoableWrapper != null) {
+                        undoableWrapper.setActive(false, null);
```

Comment:
Should `if (undoableWrapper != null) { undoableWrapper.setActive(false, null); }` be within a finally block?

Reply:
It's ok as it is.